File: Java/lab04_Builder/task_3_1/src/Engine.java
public class Engine {
    public enum Fuel {
        Petrol,
        Diesel,
        Electric
    }

    final int power;
    final int torque;
    final Fuel fuel;
    final float volume;

    public Engine(int power, int torque, Fuel fuel, float volume) {
        this.power = power;
        this.torque = torque;
        this.fuel = fuel;
        this.volume = volume;
    }

    @Override
    public String toString() {
        return "Engine{" +
                "power=" + power +
                ", torque=" + torque +
                ", fuel=" + fuel +
                ", volume=" + volume +
                '}';
    }
}

File: Cpp/lab04_Builder/task_3_1/src/engine.cpp
#include "engine.hpp"


Engine::Engine(int power, int torque, Engine::Fuel fuel, float volume)
: power(power), torque(torque), fuel(fuel), volume(volume) {}

std::ostream& operator<<(std::ostream& os, Engine const& engine)
{
  os << "Engine{power=" << engine.power
     << ", torque=" << engine.torque
     << ", fuel=";

  switch (engine.fuel)
  {
    case Engine::Fuel::Petrol:
      os << "Petrol";
      break;
    case Engine::Fuel::Diesel:
      os << "Diesel";
      break;
    case Engine::Fuel::Electric:
      os << "Electric";
      break;
  }

  os << ", volume=" << engine.volume << '}';

  return os;
}
